### PR TITLE
remote-ssh: fix Node not patching automatically

### PR DESCRIPTION
**Pull request by Enzime**
###### Description of changes

I've updated the logic to use Node 16 instead of Node 12, then just attempts to use Node from Nix if possible which matters on NixOS hosts where the bundled Node is dynamically linked and won't run.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix [hunk 1]
```diff
@@ -9,29 +9,68 @@
...
   patch = ''
-    f="$HOME/.vscode-server/bin/$COMMIT_ID/node"
-    localNodePath=''$(which node)
-    if [ -x "''$localNodePath" ]; then
-      localNodeVersion=''$(node -v)
-      if [ "\''${localNodeVersion:1:2}" = "12" ]; then
-        echo PATCH: replacing ''$f with ''$localNodePath
-        rm ''$f
-        ln -s ''$localNodePath ''$f
+    d="$HOME/.vscode-server/bin/$COMMIT_ID"
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix:19]: Can we use more descriptive variables?

Comment (Enzime) [reply pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix]: Fixed :+1:

File: pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix [hunk 2]
```diff
@@ -9,29 +9,68 @@
...
+        echo "Getting Node from Nix failed, use Local Node instead"
+        nodePath=$(which node)
+        echo "Local Node: $nodePath"
+        nodeVersion=$($nodePath -v)
+        echo "Local Node Version: $nodeVersion"
+      fi
+
+      if [ "\''${nodeVersion:1:2}" == "${nodeVersion}" ]; then
+        echo PATCH: replacing $f with $nodePath
+        rm $f
```
[the remark this answers is not in the document]
Comment (Enzime) [reply pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix]: Fixed :+1:

File: pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix [hunk 3]
```diff
@@ -9,29 +9,68 @@
...
+        ln -s $nodePath $f
```
[the remark this answers is not in the document]
Comment (Enzime) [reply pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix]: Fixed :+1:

File: pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix [hunk 4]
```diff
@@ -9,29 +9,68 @@
...
+      echo "Unsupported VS Code Node version: $nodeVersion", quitting
+      fail_with_exitcode ''${o.InstallExitCode.ServerTransferFailed}
+    fi
+
     ${lib.optionalString useLocalExtensions ''
       # Use local extensions
       if [ -d $HOME/.vscode/extensions ]; then
-        if ! test -L "$HOME/.vscode-server/extensions"; then
+        if ! [ -L $HOME/.vscode-server/extensions ]; then
+          if [ -e $HOME/.vscode-server/extensions ]; then
```
[the remark this answers is not in the document]
Comment (Enzime) [reply pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix]: I've simplified the logic by removing the symbolic check on `~/.vscode-server/extensions` :+1:

File: pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix [hunk 5]
```diff
@@ -9,29 +9,68 @@
...
+
           mkdir -p $HOME/.vscode-server
-          ln -s $HOME/.vscode/extensions $HOME/.vscode-server/
+          ln -s $HOME/.vscode/extensions $HOME/.vscode-server/extensions
         fi
       fi
     ''}
-  '';
+
+    echo "Checking $VSCH_LOGFILE''; # Don't add a trailing newline as we only matched part of a command
```
[the remark this answers is not in the document]
Comment (Enzime) [reply pkgs/applications/editors/vscode/extensions/remote-ssh/default.nix]: I've changed it to match on a full line comment `# Start the server` instead

**Comment by Enzime** [comment]
@SuperSandro2000 thanks so much for reviewing it, I've updated the PR to take into consideration your suggestions, please review it again if you have another chance

**Pull request merged**